File: src/features/typing/data/ts/number_of_islands.ts
function numIslands(grid: string[][]): number {
    const rows = grid.length;
    const cols = grid[0].length;
    const sink = (i: number, j: number): void => {
        if (i < 0 || i >= rows || j < 0 || j >= cols || grid[i][j] !== '1') {
            return;
        }
        grid[i][j] = '0';
        sink(i + 1, j);
        sink(i - 1, j);
        sink(i, j + 1);
        sink(i, j - 1);
    };
    let total = 0;
    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < cols; j++) {
            if (grid[i][j] === '1') {
                total++;
                sink(i, j);
            }
        }
    }
    return total;
}

File: src/features/typing/data/go/number_of_islands.go
package main

func numIslands(grid [][]byte) int {
	rows, cols := len(grid), len(grid[0])
	var sink func(i, j int)
	sink = func(i, j int) {
		if i < 0 || i >= rows || j < 0 || j >= cols || grid[i][j] != '1' {
			return
		}
		grid[i][j] = '0'
		sink(i+1, j)
		sink(i-1, j)
		sink(i, j+1)
		sink(i, j-1)
	}
	total := 0
	for i := 0; i < rows; i++ {
		for j := 0; j < cols; j++ {
			if grid[i][j] == '1' {
				total++
				sink(i, j)
			}
		}
	}
	return total
}
